### PR TITLE
require a password reset when too many failed password attempts are made...

### PR DESCRIPTION
**Pull request by lloyd**
... - issue #2656 

Even though there's zero UI change, I claim this is ready for merge

### CODE REVIEW

File: lib/wsapi/authenticate_user.js [hunk 1]
```diff
@@ -86,28 +88,94 @@ exports.process = function(req, res) {
...
 
 
+function incrementAuthCount(req, res, uid) {
+  var u = wsapi.forwardWritesTo;
+  var m = u.scheme === 'http' ? http : https;
+  var preq = m.request({
+    host: u.host,
+    port: u.port,
+    path: '/wsapi/increment_failed_auth_tries?' + querystring.stringify({userid: uid}),
+    method: "GET",
```
Comment (benadida) [inline lib/wsapi/authenticate_user.js:127]: should this be a POST?

File: lib/wsapi/authenticate_user.js [hunk 2]
```diff
@@ -86,28 +88,94 @@ exports.process = function(req, res) {
...
+
+function resetFailedAuthCount(req, res, uid) {
+  var u = wsapi.forwardWritesTo;
+  var m = u.scheme === 'http' ? http : https;
+  var preq = m.request({
+    host: u.host,
+    port: u.port,
+    path: '/wsapi/reset_failed_auth_tries',
+    method: "GET",
+    agent: false,
```
Comment (benadida) [inline lib/wsapi/authenticate_user.js:156]: same here, should this be a POST?

Comment (lloyd) [inline lib/wsapi/authenticate_user.js:156]: this is an internal only API call from browserid (with only DB read access) to dbwriter.  if it's a POST, we must forward cookies and CSRF tokens, which seems meaningless.  what do you think?

File: lib/wsapi/authenticate_user.js (hunk 1 above)
Comment (lloyd) [inline lib/wsapi/authenticate_user.js:127]: ditto: this is an internal only API call from browserid (with only DB read access) to dbwriter. if it's a POST, we must forward cookies and CSRF tokens, which seems meaningless. what do you think?

File: lib/wsapi/reset_failed_auth_tries.js [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+const
+db = require('../db.js'),
+wsapi = require('../wsapi.js');
+
+exports.method = 'get';
+exports.writes_db = true;
+exports.authed = true;
+exports.internal = true;
```
Comment (benadida) [inline lib/wsapi/reset_failed_auth_tries.js:12]: I'm assuming this guarantees that this call cannot be made from the user's browser. Might be good to test for this (I don't think I see any test that covers this.)

**Comment by benadida** [comment]
to @lloyd re: discussion of GET v. POST.

Longer term, I suggest sticking to standard GET/POST behavior, maybe even disabling CSRF protection on internal calls if that makes things easier. But I can live with this as is given the urgency of the fix.

File: lib/wsapi/reset_failed_auth_tries.js (hunk 3 above)
Comment (lloyd) [inline lib/wsapi/reset_failed_auth_tries.js:12]: Good point.  We had internal tests, but they did not excercise all internal apis.  update in commit 1a4572c.

**Comment by lloyd** [comment]
@benadida 

re GET v. POST on internal APIs - I agree on both counts - we should change things and make these exempt from CSRF checks where appropriate, and it shouldn't block this PR.

re testing accessibility of internal APIs, our `internal-wsapi-test.js` is updated and does this (1a4572c)

re test stability - it's fixed (205bbe9)

at present, pending an r+ from @fmarier and @benadida, I believe this is ready to merge.

**Comment by benadida** [comment]
r+ from me

**Comment by fmarier** [comment]
r+ from me too

**Comment by lloyd** [comment]
holding on this till the final FirefoxOS patches land.  @shane-tomlinson or @jedp ping me when everything needed to support FirefoxOS is done and I'll rebase and land this if I can do it safely.

**Pull request merged**